File: backend/src/erp/domain/value_objects/rut.py
"""Value object `Rut` con validación del dígito verificador chileno (módulo 11)."""
from __future__ import annotations

import re
from dataclasses import dataclass

from erp.domain.exceptions import RutInvalidoError

_RUT_RE = re.compile(r"^\d{1,8}-[\dkK]$")


def _calcular_dv(numero: int) -> str:
    suma = 0
    multiplicador = 2
    n = numero
    while n > 0:
        suma += (n % 10) * multiplicador
        n //= 10
        multiplicador = 2 if multiplicador == 7 else multiplicador + 1
    resto = 11 - (suma % 11)
    if resto == 11:
        return "0"
    if resto == 10:
        return "K"
    return str(resto)


@dataclass(frozen=True)
class Rut:
    """Representa un RUT chileno normalizado: `12345678-5` (sin puntos, con guion, dv mayúscula)."""

    valor: str

    def __post_init__(self) -> None:
        normalizado = self._normalizar(self.valor)
        if not _RUT_RE.match(normalizado):
            raise RutInvalidoError(f"RUT con formato inválido: {self.valor!r}")
        numero_str, dv = normalizado.split("-")
        if _calcular_dv(int(numero_str)) != dv:
            raise RutInvalidoError(f"Dígito verificador inválido para {self.valor!r}")
        object.__setattr__(self, "valor", normalizado)

    @staticmethod
    def _normalizar(raw: str) -> str:
        s = raw.strip().upper().replace(".", "").replace(" ", "")
        if "-" not in s and len(s) >= 2:
            s = f"{s[:-1]}-{s[-1]}"
        return s

    def __str__(self) -> str:
        return self.valor

File: frontend/src/modules/administracion/rut.ts
/**
 * Validación de RUT chileno (módulo 11).
 * Acepta formato con o sin puntos y con guión: "12.345.678-9", "12345678-9", "123456789".
 * Retorna el RUT canónico "12345678-9" o null si es inválido.
 */
export function validarRut(input: string): string | null {
  if (!input) return null;
  const cleaned = input.replace(/[^0-9kK]/g, "").toUpperCase();
  if (cleaned.length < 2 || cleaned.length > 9) return null;
  const cuerpo = cleaned.slice(0, -1);
  const dv = cleaned.slice(-1);
  if (!/^\d+$/.test(cuerpo)) return null;

  // Cálculo módulo 11
  let suma = 0;
  let multiplicador = 2;
  for (let i = cuerpo.length - 1; i >= 0; i--) {
    suma += Number(cuerpo[i]) * multiplicador;
    multiplicador = multiplicador === 7 ? 2 : multiplicador + 1;
  }
  const resto = 11 - (suma % 11);
  let dvCalc: string;
  if (resto === 11) dvCalc = "0";
  else if (resto === 10) dvCalc = "K";
  else dvCalc = String(resto);

  if (dvCalc !== dv) return null;
  return `${cuerpo}-${dv}`;
}

/** Formatea un RUT canónico añadiendo separadores de miles. */
export function formatearRut(rut: string): string {
  const m = /^(\d+)-([\dkK])$/.exec(rut);
  if (!m) return rut;
  const [, cuerpo, dv] = m;
  const conPuntos = cuerpo!.replace(/\B(?=(\d{3})+(?!\d))/g, ".");
  return `${conPuntos}-${dv!.toUpperCase()}`;
}
